### PR TITLE
Fix a bug where using groups and counts with long table names would return incorrect results.

### LINKED ISSUE #46285: Active Record GROUP_BY/COUNT calculation returns incorrect result
### Steps to reproduce
Is this intentional?
```rb
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "~> 7.0.0"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  # テーブル定義
  create_table :evaluation_template_role_scene_item_restrictions, force: true do |t|
    t.bigint :evaluation_template_sheet_item_id, null: false
    t.bigint :evaluation_template_role_scene_restriction_id, null: false
    t.timestamps
  end
end

module EvaluationTemplate
  module Role
    class SceneItemRestriction < ActiveRecord::Base
      self.table_name = 'evaluation_template_role_scene_item_restrictions'
    end
  end
end

class BugTest < Minitest::Test
  def test_association_stuff
    2.times do
      EvaluationTemplate::Role::SceneItemRestriction.create!(
        evaluation_template_role_scene_restriction_id: 1,
        evaluation_template_sheet_item_id: 2
      )
    end

    res = EvaluationTemplate::Role::SceneItemRestriction.group(:evaluation_template_role_scene_restriction_id, :evaluation_template_sheet_item_id).having("count(*) >= 2").count

    # ここが期待通り動かない
    assert_equal({ [1, 2] => 2 }, res)
  end
end
```

### Expected behavior
It should return expected count grouping by 2 columns.
```
{[1, 2]=>2}
```

### Actual behavior
```
{[2, 2]=>2}
```

### System configuration
**Rails version**: gem "activerecord", "~> 7.0.0"

**Ruby version**: ruby 3.0.2p107

### PR DESCRIPTION
**Pull request by Dooor**
### Motivation / Background

Calculation results are buggy when using group by and count when table names are long.

Fixed https://github.com/rails/rails/issues/46285

### Detail

Fixed a bug that caused the alias name of "group by" to be too long and the first half of the name would be the same in both cases if it was cut by max identifier length.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [ ] CI is passing.

### CODE REVIEW

File: activerecord/CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,8 @@
-*   Fix encryption of column default values.
+*   Fix a bug where using groups and counts with long table names would return incorrect results.
+
+    *Shota Toguchi*
```
Comment (kamipo) [inline activerecord/CHANGELOG.md:3]: I know @Dooor and @yono are a member of Today's "How to Rails Contributing" meeting.
So how about adding co-authored-by for all you guys.

File: activerecord/CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,8 @@
...
+
+* Fix encryption of column default values.
```
Comment (kamipo) [inline activerecord/CHANGELOG.md:5]:
```suggestion
*    Fix encryption of column default values.
```

File: activerecord/CHANGELOG.md (hunk 1 above)
Comment (Dooor) [reply activerecord/CHANGELOG.md]: Thanks for comment. I added @yono to co-authored-by in commit message and CHANGELOG.md.

File: activerecord/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+*   Fix a bug where using groups and counts with long table names would return incorrect results.
+
+    *Shota Toguchi, Yusaku Ono*
```
Comment (kamipo) [inline activerecord/CHANGELOG.md:3]:
```suggestion
    *Shota Toguchi*, *Yusaku Ono*
```

Comment (kamipo) [reply activerecord/CHANGELOG.md:3]: Fixed by 4136145819abfdb3fe4016cf767aa1eb346c7bae.

**Comment by kamipo** [comment]
Backported to 7-0-stable a6d0f80d3af575482d826301d947e43fc3c5c649.

**Pull request merged**